File: samples/litert/image_generation/android/app/src/main/java/com/google/ai/edge/samples/imagegeneration/QwenTokenizer.kt
// Copyright 2026 The Google AI Edge Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// =============================================================================

// Qwen3 tokenizer (byte-level BPE) in pure Kotlin — enough to feed the Bonsai
// text encoder on device. Port of the Swift implementation (../ios/Sources/
// QwenTokenizer.swift); loads vocab.json + merges.txt from the HF repo's
// tokenizer/ folder. The chat template is applied structurally: the graph
// input is [<|im_start|>] + BPE("user\n" + prompt) + a fixed assistant
// suffix, which reproduces transformers' apply_chat_template(...,
// enable_thinking=False) exactly. "user\n" must be BPE'd TOGETHER with the
// prompt (a prompt starting with whitespace merges across that boundary).
// Verified token-exact against the Python tokenizer on the same 26-case
// golden set as the Swift port (JVM test QwenTokenizerTest).
//
// Not handled (irrelevant for an image-prompt box): special-token strings
// typed literally inside the prompt are tokenized as plain text.

package com.google.ai.edge.samples.imagegeneration

import org.json.JSONObject
import java.io.InputStream

class QwenTokenizer(vocabJson: InputStream, mergesTxt: InputStream) {
    companion object {
        const val SEQ_LEN = 256
        const val PAD_ID = 151643        // <|endoftext|>
        const val IM_START_ID = 151644   // <|im_start|>
        // "<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\n"
        val SUFFIX_IDS = intArrayOf(151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271)

        // Qwen2 pre-tokenization pattern, verbatim from tokenizer.json.
        // java.util.regex: (?i:...) group works; \p{L}/\p{N} need UNICODE flags.
        private val PRETOKEN = Regex(
            "(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}|" +
                " ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+",
            setOf(RegexOption.UNIX_LINES)
        )

        // GPT-2 bytes-to-unicode: printable bytes map to themselves, the rest
        // to U+0100... so every byte is a distinct printable char and ' '
        // never appears inside a symbol (making "left right" rank keys safe).
        private val BYTE_CHAR: CharArray = run {
            val bs = ((33..126) + (161..172) + (174..255)).toMutableList()
            val cs = bs.toMutableList()
            var n = 0
            for (b in 0..255) if (b !in bs) {
                bs.add(b); cs.add(256 + n); n++
            }
            val table = CharArray(256)
            for (i in bs.indices) table[bs[i]] = cs[i].toChar()
            table
        }
    }

    private val vocab = HashMap<String, Int>(160_000)
    private val ranks = HashMap<String, Int>(160_000)
    private val cache = HashMap<String, IntArray>()

    init {
        val jo = JSONObject(vocabJson.bufferedReader().readText())
        for (key in jo.keys()) vocab[key] = jo.getInt(key)
        mergesTxt.bufferedReader().forEachLine { line ->
            if (line.isNotEmpty() && !line.startsWith("#")) ranks[line] = ranks.size
        }
    }

    class Encoded(val ids: IntArray, val mask: IntArray, val promptTokenCount: Int)

    /** Full graph input for one user prompt: template + pad to 256, right-pad
     *  mask. Over-long prompts are truncated so the assistant suffix survives. */
    fun encodePrompt(prompt: String): Encoded {
        var body = encode("user\n" + prompt)
        val maxBody = SEQ_LEN - 1 - SUFFIX_IDS.size
        if (body.size > maxBody) body = body.copyOf(maxBody)
        val real = 1 + body.size + SUFFIX_IDS.size
        val ids = IntArray(SEQ_LEN) { PAD_ID }
        val mask = IntArray(SEQ_LEN)
        ids[0] = IM_START_ID
        body.copyInto(ids, 1)
        SUFFIX_IDS.copyInto(ids, 1 + body.size)
        for (i in 0 until real) mask[i] = 1
        return Encoded(ids, mask, body.size)
    }

    /** Byte-level BPE of plain text (no special-token splitting). */
    fun encode(text: String): IntArray {
        val nfc = java.text.Normalizer.normalize(text, java.text.Normalizer.Form.NFC)
        val out = ArrayList<Int>(nfc.length / 3 + 8)
        for (m in PRETOKEN.findAll(nfc)) {
            if (m.value.isEmpty()) continue
            bpe(m.value).forEach { out.add(it) }
        }
        return out.toIntArray()
    }

    private fun bpe(pretoken: String): IntArray {
        cache[pretoken]?.let { return it }
        var word = pretoken.toByteArray(Charsets.UTF_8).map {
            BYTE_CHAR[it.toInt() and 0xFF].toString()
        }
        while (word.size > 1) {
            var best = Int.MAX_VALUE
            var at = -1
            for (i in 0 until word.size - 1) {
                val r = ranks[word[i] + " " + word[i + 1]] ?: continue
                if (r < best) { best = r; at = i }
            }
            if (at < 0) break
            val a = word[at]
            val b = word[at + 1]
            val merged = ArrayList<String>(word.size)
            var i = 0
            while (i < word.size) {
                if (i < word.size - 1 && word[i] == a && word[i + 1] == b) {
                    merged.add(a + b); i += 2
                } else {
                    merged.add(word[i]); i += 1
                }
            }
            word = merged
        }
        // Byte-level alphabet: every symbol/merge result exists in the vocab.
        val ids = word.mapNotNull { vocab[it] }.toIntArray()
        cache[pretoken] = ids
        return ids
    }
}

File: samples/litert/image_generation/macos/Sources/QwenTokenizer.swift
// Copyright 2026 Daisuke Majima. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// =============================================================================

// Qwen3 tokenizer (byte-level BPE) in pure Swift — enough to feed the Bonsai
// text encoder on device. Loads vocab.json + merges.txt (the tokenizer/ folder
// of the HF repo). The chat template is applied structurally: the graph input
// is [<|im_start|>] + BPE("user\n" + prompt) + a fixed assistant suffix, which
// reproduces transformers' apply_chat_template(..., enable_thinking=False)
// exactly — the <|im_end|> added token forces a pre-tokenization split, so the
// suffix ids are prompt-independent constants, while "user\n" must be BPE'd
// TOGETHER with the prompt (a prompt starting with whitespace merges across
// that boundary). Verified against the Python tokenizer on a 26-case golden
// set (ASCII/CJK/emoji/NFD/whitespace edges) before shipping.
//
// Not handled (irrelevant for an image-prompt box): special-token strings
// typed literally inside the prompt are tokenized as plain text.

import Foundation

final class QwenTokenizer {
    static let seqLen = 256
    static let padID: Int32 = 151643        // <|endoftext|>
    static let imStartID: Int32 = 151644    // <|im_start|>
    // "<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\n"
    static let suffixIDs: [Int32] = [151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271]

    private let vocab: [String: Int32]
    private let ranks: [String: Int]
    private let regex: NSRegularExpression
    private var cache: [String: [Int32]] = [:]

    // Qwen2 pre-tokenization pattern, verbatim from tokenizer.json.
    private static let pattern =
        "(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}|" +
        " ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+"

    // GPT-2 bytes-to-unicode: printable bytes map to themselves, the rest to
    // U+0100... so every byte is a distinct printable scalar and " " never
    // appears inside a symbol (which makes "left right" rank keys unambiguous).
    private static let byteChar: [Character] = {
        var bs = Array(33...126) + Array(161...172) + Array(174...255)
        var cs = bs
        var n = 0
        for b in 0...255 where !bs.contains(b) {
            bs.append(b)
            cs.append(256 + n)
            n += 1
        }
        var table = [Character](repeating: " ", count: 256)
        for (b, c) in zip(bs, cs) { table[b] = Character(UnicodeScalar(c)!) }
        return table
    }()

    init(vocabURL: URL, mergesURL: URL) throws {
        let raw = try JSONSerialization.jsonObject(with: Data(contentsOf: vocabURL))
        guard let dict = raw as? [String: Int] else {
            throw NSError(domain: "QwenTokenizer", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "vocab.json is not {token: id}"])
        }
        var v: [String: Int32] = [:]
        v.reserveCapacity(dict.count)
        for (k, id) in dict { v[k] = Int32(id) }
        vocab = v

        var r: [String: Int] = [:]
        let merges = try String(contentsOf: mergesURL, encoding: .utf8)
        var rank = 0
        merges.enumerateLines { line, _ in
            if line.hasPrefix("#") || line.isEmpty { return }
            r[line] = rank
            rank += 1
        }
        ranks = r
        regex = try NSRegularExpression(pattern: Self.pattern)
    }

    struct Encoded {
        let ids: [Int32]
        let mask: [Int32]
        let promptTokenCount: Int
    }

    /// Full graph input for one user prompt: template + pad to 256, right-pad
    /// mask. Over-long prompts are truncated so the assistant suffix survives.
    func encodePrompt(_ prompt: String) -> Encoded {
        var body = encode("user\n" + prompt)
        let maxBody = Self.seqLen - 1 - Self.suffixIDs.count
        if body.count > maxBody { body = Array(body.prefix(maxBody)) }
        var ids = [Self.imStartID] + body + Self.suffixIDs
        let real = ids.count
        ids.append(contentsOf: [Int32](repeating: Self.padID, count: Self.seqLen - real))
        let mask = [Int32](repeating: 1, count: real)
                 + [Int32](repeating: 0, count: Self.seqLen - real)
        return Encoded(ids: ids, mask: mask, promptTokenCount: body.count)
    }

    /// Byte-level BPE of plain text (no special-token splitting).
    func encode(_ text: String) -> [Int32] {
        let nfc = text.precomposedStringWithCanonicalMapping
        let ns = nfc as NSString
        var out: [Int32] = []
        let matches = regex.matches(in: nfc, range: NSRange(location: 0, length: ns.length))
        for m in matches {
            out.append(contentsOf: bpe(ns.substring(with: m.range)))
        }
        return out
    }

    private func bpe(_ pretoken: String) -> [Int32] {
        if let hit = cache[pretoken] { return hit }
        var word: [String] = pretoken.utf8.map { String(Self.byteChar[Int($0)]) }
        while word.count > 1 {
            var best = Int.max
            var at = -1
            for i in 0..<(word.count - 1) {
                if let r = ranks[word[i] + " " + word[i + 1]], r < best {
                    best = r
                    at = i
                }
            }
            if at < 0 { break }
            let a = word[at], b = word[at + 1]
            var merged: [String] = []
            merged.reserveCapacity(word.count)
            var i = 0
            while i < word.count {
                if i < word.count - 1, word[i] == a, word[i + 1] == b {
                    merged.append(a + b)
                    i += 2
                } else {
                    merged.append(word[i])
                    i += 1
                }
            }
            word = merged
        }
        // Byte-level alphabet: every symbol/merge result exists in the vocab.
        let ids = word.compactMap { vocab[$0] }
        cache[pretoken] = ids
        return ids
    }
}
